Sprint 11

Georgy-hook opened the pull request.

Georgy-hook committed:
- Сode corrected as recommended
- Added ProgressHUD

Georgy-hook committed:
- Race conditions fixed

Georgy-hook committed:
- Added ProfileService
- Added ProfileImageService
- Added NotificationCenter observer (ProfileViewController)

Georgy-hook committed:
- Сhanged network interaction of classes OAuth2Service, ProfileService, ProfileImageService
- Added generic method objectTask
- Delete method data
- Added class ErrorAlertService
- Added show alerts for SplashScreenViewController

Georgy-hook committed:
- All view controllers are layout in code

amardanov commented on ImageFeed/Presentation/Auth/WebViewViewController.swift:
```diff
@@ -30,42 +55,35 @@ final class WebViewViewController: UIViewController {
             URLQueryItem(name: "scope", value: AccessScope)
         ]
         let url = urlComponents.url!
```
Лучше такое обернуть в `guard`, чтобы не падало приложение, если вдруг `urlComponents` неправильно собирается

amardanov commented on ImageFeed/Presentation/ImagesListCell.swift:
```diff
@@ -8,34 +8,95 @@
-        configure()
+
+    //MARK: - UI Elements
+    let imageCell: UIImageView = {
```
все UI свойства лучше сделать private, а их конфигурацию вынести в метод configure

amardanov commented on ImageFeed/Presentation/SplashScreenViewController.swift:
```diff
@@ -6,22 +6,37 @@
 //
 
 import UIKit
+import ProgressHUD
```
Здесь надо убрать этот импорт и использовать `UIBlockingProgressHUD`

amardanov commented on ImageFeed/Presentation/SplashScreenViewController.swift:
```diff
@@ -6,22 +6,37 @@
 
 final class SplashScreenViewController: UIViewController {
+    //MARK: - SplashScreen ImageView
+    let splashScreenImageView:UIImageView = {
```
private

amardanov commented on ImageFeed/Presentation/SplashScreenViewController.swift:
```diff
@@ -34,50 +49,84 @@ final class SplashScreenViewController: UIViewController {
+    }
+    private func fetchProfileImageURL(username: String, token: String){
+        profileImageService.fetchProfileImageURL(username: username , token){ [weak self] result in
+            //      guard let self = self else { return }
```
`[weak self]` тут не нужен и надо убрать коммент код

amardanov commented on ImageFeed/Services/OAuth2Service.swift:
```diff
@@ -9,47 +9,49 @@ import Foundation
         _ code: String,
         completion: @escaping (Result<String, Error>) -> Void
     ) {
+        assert(Thread.isMainThread)
```
`assert` стоит использовать только в тестах, а отсюда убрать и лучше делать не  `if lastCode == code { return }`, а `guard lastCode != code else { return }`

amardanov commented on ImageFeed/Services/ProfileImageService.swift:
```diff
@@ -0,0 +1,61 @@
+//
+
+import UIKit
+import Kingfisher
```
Это здесь не надо

amardanov commented on ImageFeed/Services/ProfileImageService.swift:
```diff
@@ -0,0 +1,61 @@
+                object: self,
+                userInfo: ["URL": avatarURL])
+        
+        assert(Thread.isMainThread)
```
Тоже самое про Гард и ассерты

amardanov commented on ImageFeed/Services/ProfileImageService.swift:
```diff
@@ -0,0 +1,61 @@
+        request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
+        return request
+    }
+    private func loadImage(with URL:String){
```
не Нужный метод кажется

amardanov commented on ImageFeed/Services/ProfileService.swift:
```diff
@@ -0,0 +1,53 @@
+    private(set) var profile: Profile?
+    
+    func fetchProfile(_ token: String, completion: @escaping (Result<Profile, Error>) -> Void){
+        assert(Thread.isMainThread)
```
Тоже самое про ассерт и гард

Georgy-hook committed:
- fixed bug when pressing exit button
- fixed decoding error if last name = nil
- added self. to SplashScreenViewController
- fixed access rights
- deleted unused method
- deleted unused imports

Georgy-hook committed:
- add another self
